[clang] Fix ASTUnit working directory handling

hamishknight opened the pull request:
Fix a couple of issues with the handling of the current working directory in ASTUnit:

- Use `createPhysicalFileSystem` instead of `getRealFileSystem` to avoid affecting the process' current working directory, and set it at the top of `ASTUnit::LoadFromCommandLine` such that the driver used for argument parsing and the ASTUnit share the same VFS. This ensures that '-working-directory' correctly sets the VFS working directory in addition to the FileManager working directory.
- Ensure we preserve the FileSystemOptions set on the FileManager when re-creating it (as `ASTUnit::Reparse` will clear the currently set FileManager).

rdar://110697657

hamishknight commented:
@swift-ci please test

bnbarham commented on clang/unittests/Frontend/ReparseWorkingDirTest.cpp:
```diff
@@ -0,0 +1,118 @@
+
+  std::unique_ptr<ASTUnit> ParseAST(StringRef EntryFile) {
+    PCHContainerOpts = std::make_shared<PCHContainerOperations>();
+    std::shared_ptr<CompilerInvocation> CI(new CompilerInvocation);
```
`make_shared`?

bnbarham commented on clang/unittests/Frontend/ReparseWorkingDirTest.cpp:
```diff
@@ -0,0 +1,118 @@
+
+TEST_F(ReparseWorkingDirTest, ReparseWorkingDir) {
+  SmallString<64> WorkingDir;
+#ifdef _WIN32
+  WorkingDir = "C:\\";
+#else
+  WorkingDir = "/";
+#endif
```
`InMemoryVFS` is being used anyway, does it matter what the root is?

hamishknight replied:
Possibly not, I was basing this off this comment:

https://github.com/apple/llvm-project/blob/e1bf2eacb1fc61cf875441e462dbc376e620fd22/clang/unittests/Frontend/PCHPreambleTest.cpp#L64-L70

which makes it sound like it needs to be valid for the platform. I can try drop it to see if it's fine with just `/` tho

bnbarham replied:
Hmmm. How about just `//root` like the `VirtualFileSystemTest.cpp` tests then?

hamishknight replied:
Yeah, that works

bnbarham commented on clang/unittests/Frontend/ReparseWorkingDirTest.cpp:
```diff
@@ -0,0 +1,118 @@
+static int foo() { return 0; }
+)h");
+
+  // Parse with header file provided as unsaved file, which does not exist on
+  // disk.
```
Is this comment copied from somewhere? Technically it's true because we're using the in-memory VFS, but seems a bit weird.

hamishknight replied:
Oops yes, I copied this test case from another file, I'll update the comment.

bnbarham reviewed (approved):
Thanks for tracking this down @hamishknight!

benlangmuir commented on clang/lib/Frontend/ASTUnit.cpp:
```diff
@@ -1752,6 +1752,12 @@ ASTUnit *ASTUnit::LoadFromCommandLine(
+  // If no VFS was provided, create one that tracks the physical file system.
+  // If '-working-directory' was passed as an argument, 'createInvocation' will
+  // set this as the current working directory of the VFS.
+  if (!VFS)
```
Why is this done here instead of above the first use of VFS to create a FileManager above?

hamishknight replied:
Hmm, could you elaborate? This is the first use of `VFS` in the function. Or do you mean why are we doing it here instead of where we were previously doing it (immediately before creating the FileManager)? We're doing it here so that it can be used by `createInvocation`, such that `-working-directory` will set the working directory of the VFS.

benlangmuir replied:
Sorry, I just confused myself when looking at the rest of the file and lost track of where exactly this change lined up.

benlangmuir reviewed (commented):
These changes should go upstream to llvm.org

hamishknight commented:
> These changes should go upstream to llvm.org

Yup, that's the plan, we're cherry-picking downstream first as we want to get this into 5.9 fairly soon

bnbarham commented:
@swift-ci please test Windows platform

benlangmuir commented on clang/unittests/Frontend/ASTUnitTest.cpp:
```diff
@@ -178,4 +178,36 @@ TEST_F(ASTUnitTest, LoadFromCommandLineEarlyError) {
+  // VFS working directory.
+  const auto &FM = AST->getFileManager();
+  const auto &VFS = FM.getVirtualFileSystem();
+  ASSERT_STREQ(VFS.getCurrentWorkingDirectory()->data(), WorkingDir.c_str());
```
Why STREQ instead of EQ on StringRef?

hamishknight replied:
Ah nice, didn't realize you could do that

hamishknight committed:
[ASTUnit] Use `createPhysicalFileSystem` if no VFS was provided

Use `createPhysicalFileSystem` instead of `getRealFileSystem` to
avoid affecting the process' current working directory, and set it
at the top of `ASTUnit::LoadFromCommandLine` such that the driver
used for argument parsing and the ASTUnit share the same VFS. This
ensures that '-working-directory' correctly sets the VFS working
directory in addition to the FileManager working directory.

hamishknight committed:
[ASTUnit] Ensure we set the invocation before creating the FileManager

Ensure we preserve the FileSystemOptions set on the FileManager when
re-creating it (as `ASTUnit::Reparse` will clear the currently set
FileManager).

rdar://110697657

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test macOS

hamishknight commented:
@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1855

@swift-ci please test macOS

hamishknight commented:
@swift-ci please test macOS

hamishknight commented:
Upstream patch: https://reviews.llvm.org/D154134
